quickjs.c

```diff
@@ -50132,3 +50417,168 @@ void JS_AddPerformance(JSContext *ctx)
                            JS_PROP_ENUMERABLE);
     JS_FreeValue(ctx, performance);
 }
+
+/* Poly IC */
+
+static force_inline uint32_t get_index_hash(JSAtom atom, int hash_bits)
+{
+    return (atom * 0x9e370001) >> (32 - hash_bits);
+}
+
+InlineCache *init_ic(JSContext *ctx)
+{
+    InlineCache *ic;
+    ic = js_malloc(ctx, sizeof(InlineCache));
+    if (unlikely(!ic))
+        goto fail;
+    ic->count = 0;
+    ic->hash_bits = 2;
+    ic->capacity = 1 << ic->hash_bits;
+    ic->ctx = ctx;
+    ic->hash = js_malloc(ctx, sizeof(ic->hash[0]) * ic->capacity);
+    if (unlikely(!ic->hash))
+        goto fail;
+    memset(ic->hash, 0, sizeof(ic->hash[0]) * ic->capacity);
+    ic->cache = NULL;
+    ic->updated = FALSE;
+    ic->updated_offset = 0;
+    return ic;
+fail:
+    return NULL;
+}
+
+int rebuild_ic(InlineCache *ic)
+{
+    uint32_t i, count;
+    InlineCacheHashSlot *ch;
+    if (ic->count == 0)
+        goto end;
+    count = 0;
+    ic->cache = js_malloc(ic->ctx, sizeof(InlineCacheRingSlot) * ic->count);
+    if (unlikely(!ic->cache))
+        goto fail;
+    memset(ic->cache, 0, sizeof(InlineCacheRingSlot) * ic->count);
+    for (i = 0; i < ic->capacity; i++) {
+        for (ch = ic->hash[i]; ch != NULL; ch = ch->next) {
+            ch->index = count++;
+            ic->cache[ch->index].atom = JS_DupAtom(ic->ctx, ch->atom);
+            ic->cache[ch->index].index = 0;
+        }
+    }
+end:
+    return 0;
+fail:
+    return -1;
+}
+
+int resize_ic_hash(InlineCache *ic)
+{
+    uint32_t new_capacity, i, h;
+    InlineCacheHashSlot *ch, *ch_next;
+    InlineCacheHashSlot **new_hash;
+    ic->hash_bits += 1;
+    new_capacity = 1 << ic->hash_bits;
+    new_hash = js_malloc(ic->ctx, sizeof(ic->hash[0]) * new_capacity);
+    if (unlikely(!new_hash))
+        goto fail;
+    memset(new_hash, 0, sizeof(ic->hash[0]) * new_capacity);
```

Comment:
```suggestion
    new_hash = js_mallocz(ic->ctx, sizeof(ic->hash[0]) * new_capacity);
    if (unlikely(!new_hash))
        goto fail;
```
Note that `ic->hash_bits` has still been mutated even though we don't grow the array on error. That's probably bad.